### ISSUE TITLE
Allow deferring of error highlighting

### ISSUE THREAD

**Issue by lilred** [atom-community/linter#952]
I have an issue where the linter is too fast! Sometimes I'm writing code and error messages keep flashing before my eyes even as I'm writing the code. It'd be nice if there was a way to delay the appearance of error messages, say by a half second. (Visual Studio does this, A+ UX)

**Comment by steelbrain** [comment]
We are using Core's `onDidStopChanging` timeout, if you don't like it, just disable `lintOnFly` from settings. You'll then get issues on save.

**Comment by lilred** [comment]
Is this configurable on a per-language or per-plugin basis?

**Comment by steelbrain** [comment]
Not really. By the nature of your issue, I don't think it's language or provider bind.

**Comment by lilred** [comment]
What if a plug-in wants to offer a consistent user experience to all users? It shouldn't have to clobber global settings.

**Comment by Arcanemagus** [comment]
@lilred the root issue is that Atom's `onDidStopChanging` fires _way_ too soon for many people (myself included). It reportedly works fine for a lot of people though. Fixing the issue there is the proper solution as that would fix _everything_ that depends on this functionality, including `linter`.

**Comment by lilred** [comment]
@Arcanemagus is there an existing issue in the Atom issue tracker I can reference?

**Comment by Arcanemagus** [comment]
Actually, surprisingly enough not that I can see from a quick search :open_mouth:.

**Comment by lilred** [comment]
Let's open one! atom/atom?

**Comment by lilred** [comment]
Actually, I don't know enough about the issues at play to write up something accurate and thorough - would you mind doing it?

**Comment by Arcanemagus** [comment]
I'm looking over the discussions around the Atom forum and I think we might be trying to use this in a manner it might not be intended for:
```
# Essential: Invoke `callback` when the buffer's contents change. It is
  # emit asynchronously 300ms after the last buffer change. This is a good place
  # to handle changes to the buffer without compromising typing performance.
```
It's quite likely that it is doing exactly as it states and firing off after 300 ms. 300ms is a rather tiny delay when the user is in the middle of writing code, maybe we want to add an extra delay to that, although I'm not sure what a "good" value would be. Thoughts @steelbrain?

**Comment by steelbrain** [comment]
Well. Here's an idea, we could introduce a config that defaults to 300 ms and use the new debounce helper I introduced and use `onDidChange` instead of `onDidStopChanging`.

**Comment by lilred** [comment]
Can you link to the debounce helper in question?

**Comment by steelbrain** [comment]
https://github.com/atom-community/linter/commit/1330cca166a22845ce6d06480c726849ea11f32a

**Comment by lilred** [comment]
Lovely, thanks.

**Comment by Arcanemagus** [comment]
That sounds like a great idea to me.

**Comment by steelbrain** [comment]
It's super easy so I'm working on a PR.

**Comment by lilred** [comment]
I was thinking that it would be nice to have this be configurable on a per-plugin basis, i.e.

>What if a plug-in wants to offer a consistent user experience to all users?

Is this in the cards at all?

**Comment by steelbrain** [comment]
Nope buddy, we can't do this per-plugin. It would introduce unnecessary complexity :)

**Comment by lilred** [comment]
I imagine because there's no existing per-plugin config object?

**Comment by steelbrain** [comment]
Not because of that, but because they are just providers and we trigger them ourselves and we don't even know names of many of them. To make it work per-provider we'll have to enforce the naming requirement which will break a lot of existing providers. Not just that but currently we have only one linting lock, then we'll have one per linter. We'll have to trigger them all at different times which means the size of the bottom panel will expand and shrink at random times which can be quite annoying and we will have to dispatch all of them separately. Most of the linter registry would need to be rewritten for that.

**Comment by steelbrain** [comment]
@lilred It's been implemented in #955
